### ISSUE TITLE
Illegal instruction popcnt64

### ISSUE THREAD

**Issue by Raulshc** [kcat/openal-soft#324]
Hi, I have compiled and tested the current version of the library and I have troubles in execution time. Specifically, there is an error of _illegal instruction_ in `alc.cpp`  line 2343 with the intrinsic __popcnt64. This instruction was added in Nehalem architecture for intel processors and in K10 (Barcelona stepping) for AMD processors (see [POPCNT and LZCNT](https://en.wikipedia.org/wiki/SSE4#POPCNT_and_LZCNT) and [__popcnt](https://docs.microsoft.com/en-us/cpp/intrinsics/popcnt16-popcnt-popcnt64?view=vs-2019)). Processors prior to these architectures will not be able to execute the library, i.e. pentium, core 2 duo, quad, athlon, athlon64 ... with the current source code.

**Comment by kcat** [comment]
Hmm. Is there a way to test if a given intrinsic is supported by the compile-target processor? I would have thought that MSVC generates fallback code if the current target doesn't support the instruction, like GCC does with `__builtin_popcount[ll]`.

**Comment by iSLC** [comment]
I suppose the same thing achieved with `CheckFileOffsetBits.cmake` and `CheckFileOffsetBits.c` can be done for `popcnt` as well.

OpenCV seems to have code to detect it [cmake/checks/cpu_popcnt.cpp](https://github.com/opencv/opencv/blob/master/cmake/checks/cpu_popcnt.cpp)

On GCC you basically need to look for the presence of `__POPCNT__` macro.

**Comment by kcat** [comment]
GCC automatically handles it already. `__builtin_popcnt[ll]` will use the available opcodes when the target CPU supports them, otherwise inlines alternative code that produces the same result. MSVC, though, seems to generate the opcodes regardless of the target.

Unfortunately that OpenCV check doesn't look like it will work here, because `__popcnt[64]` will compile regardless of the target's capabilities. Attempting to run it (to check for a crash) would only say if the host system supports it or not, not the target.

**Comment by kcat** [comment]
Should be fixed now with commit e3f5bd37e632229fd7e9e849911578e58e7ccbb5.

**Comment by Raulshc** [comment]
> Should be fixed now with commit e3f5bd3.

Yes, I confirm that this problem has been fixed.

> Hmm. Is there a way to test if a given intrinsic is supported by the compile-target processor? I would have thought that MSVC generates fallback code if the current target doesn't support the instruction, like GCC does with `__builtin_popcount[ll]`.

We could do a test in cmake to detect whether the popcnt instruction is in the target processor via cpuid instruction, calling it with eax=0x00000001 and checking bit 23 of CPUInfo [2] (ecx). A sample code:
```
#include <intrin.h>
#include <stdio.h>

int main()
{
	int regs[4], out;

	__cpuid(regs, 1);
	out = ((regs[2] & 0x00800000)>>23);

	if (out)
		printf("The processor has popcnt");
	else
		printf("The processor hasn't popcnt");

}
```

**Comment by kcat** [comment]
> We could do a test in cmake to detect whether the popcnt instruction is in the target processor via cpuid instruction, calling it with eax=0x00000001 and checking bit 23 of CPUInfo [2] (ecx).

That code would need to be executed to test, which would end up testing the compiling machine for support instead of the target CPU.
